### PR TITLE
[FLINK-28928][tests] Add E2E test for hybrid shuffle mode.

### PR DESCRIPTION
**Pull request by reswqa**
## What is the purpose of the change

*Add E2E test for hybrid shuffle mode*

## Brief change log

  - *Introduce HyrbidShuffleITCase*

## Verifying this change

This change introduce a new IT test.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

File: flink-tests/src/test/java/org/apache/flink/test/runtime/HybridShuffleITCase.java [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+        checkAllDataReceived(numRecordsToSend);
+    }
+
+    @Test
+    void testHybridFullExchangesRestart() throws Exception {
+        final int numRecordsToSend = 10;
+        Configuration configuration = getConfiguration();
+        configuration.set(
+                ExecutionOptions.BATCH_SHUFFLE_MODE, BatchShuffleMode.ALL_EXCHANGES_HYBRID_FULL);
+        configuration.set(NettyShuffleEnvironmentOptions.NETWORK_REQUEST_BACKOFF_MAX, 100);
```
Comment (xintongsong) [inline flink-tests/src/test/java/org/apache/flink/test/runtime/HybridShuffleITCase.java:102]: We can move this to `getConfiguration()`.

File: flink-tests/src/test/java/org/apache/flink/test/runtime/HybridShuffleITCase.java [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+            }
+            throw new RuntimeException("expected exception.");
+        }
+
+        @Override
+        public void invoke(String value, Context context) throws Exception {
+            NUM_RECEIVED_RECORDS[getRuntimeContext().getIndexOfThisSubtask()]++;
+            assertThat(value).isEqualTo(RECORD);
+        }
+    }
```
Comment (xintongsong) [inline flink-tests/src/test/java/org/apache/flink/test/runtime/HybridShuffleITCase.java:202]: These classes seem duplications of those in `BlockingShuffleITCase`. We may introduce a `ShuffleITCaseBase` for deduplication.

Comment (xintongsong) [reply flink-tests/src/test/java/org/apache/flink/test/runtime/HybridShuffleITCase.java:202]: And can `createJobGraph` be deduplicated as well?

File: flink-end-to-end-tests/flink-batch-sql-test/src/main/java/org/apache/flink/sql/tests/BatchSQLTestProgram.java [hunk 3]
```diff
@@ -55,9 +57,10 @@ public static void main(String[] args) throws Exception {
         ParameterTool params = ParameterTool.fromArgs(args);
         String outputPath = params.getRequired("outputPath");
         String sqlStatement = params.getRequired("sqlStatement");
-
+        String shuffleMode = params.getRequired("shuffleMode");
         TableEnvironment tEnv = TableEnvironment.create(EnvironmentSettings.inBatchMode());
-
+        tEnv.getConfig()
+                .set(ExecutionOptions.BATCH_SHUFFLE_MODE, BatchShuffleMode.valueOf(shuffleMode));
```
Comment (xintongsong) [inline flink-end-to-end-tests/flink-batch-sql-test/src/main/java/org/apache/flink/sql/tests/BatchSQLTestProgram.java:63]: Let's decouple the program argument from the `BATCH_SHUFFLE_MODE` values.

File: flink-tests/src/test/java/org/apache/flink/test/runtime/HybridShuffleITCase.java (hunk 1 above)
Comment (reswqa) [reply flink-tests/src/test/java/org/apache/flink/test/runtime/HybridShuffleITCase.java:102]: Moved.

File: flink-tests/src/test/java/org/apache/flink/test/runtime/HybridShuffleITCase.java (hunk 2 above)
Comment (reswqa) [reply flink-tests/src/test/java/org/apache/flink/test/runtime/HybridShuffleITCase.java:202]: Abstract common logical to `BatchShuffleITCaseBase`.

File: flink-end-to-end-tests/flink-batch-sql-test/src/main/java/org/apache/flink/sql/tests/BatchSQLTestProgram.java (hunk 3 above)
Comment (reswqa) [reply flink-end-to-end-tests/flink-batch-sql-test/src/main/java/org/apache/flink/sql/tests/BatchSQLTestProgram.java]: done.

**Comment by reswqa** [comment]
@xintongsong Thanks for the review, I have updated this pr according to your comments, please take a look again~

**Review by xintongsong — APPROVED** [review]

**Comment by xintongsong** [comment]
@flinkbot run azure

**Comment by xintongsong** [comment]
@flinkbot run azure

**Pull request closed without merging**